Source/WebCore/Modules/applepay/PaymentSession.cpp

```diff
@@ -52,23 +52,27 @@ ExceptionOr<void> PaymentSession::canCreateSession(Document& document)
     if (!PermissionsPolicy::isFeatureEnabled(PermissionsPolicy::Feature::Payment, document))
         return Exception { ExceptionCode::SecurityError, "Third-party iframes are not allowed to request payments unless explicitly allowed via Feature-Policy (payment)"_s };
 
-    if (!document.frame())
+    RefPtr<Frame> currentFrame = document.frame();
+    if (!currentFrame)
         return Exception { ExceptionCode::InvalidAccessError, "Trying to start an Apple Pay session from an inactive document."_s };
 
-    if (!isSecure(*document.loader()))
+    if (!currentFrame->frameCanCreatePaymentSession())
         return Exception { ExceptionCode::InvalidAccessError, "Trying to start an Apple Pay session from an insecure document."_s };
 
-    RefPtr mainFrameDocument = document.protectedMainFrameDocument();
-    if (!mainFrameDocument) {
-        LOG_ONCE(SiteIsolation, "Unable to properly calculate PaymentSession::canCreateSession() without access to the main frame document ");
-        return Exception { ExceptionCode::InvalidAccessError, "Trying to start an Apple Pay session from a site isolated iframe"_s };
-    }
-
     if (!document.isTopDocument()) {
-        for (RefPtr ancestorDocument = document.parentDocument(); ancestorDocument != mainFrameDocument.get(); ancestorDocument = ancestorDocument->parentDocument()) {
-            if (!isSecure(*ancestorDocument->loader()))
+        do {
+            RefPtr parent = currentFrame->tree().parent();
+            if (!parent)
+                return Exception { ExceptionCode::InvalidAccessError, "Trying to start an Apple Pay session from a document in an unparented frame"_s };
```

Comment:
This will eventually be hit as we traverse up the tree.  I think we'll need to move the if (!document.isTopDocument()) check into the do loop if we want this.

Reply:
Fixing, but handling differently thought.